datastreams/windower/sliding.go

```diff
@@ -0,0 +1,153 @@
+package windower
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/elastiflow/pipelines/datastreams/internal/partition"
+	"github.com/elastiflow/pipelines/datastreams/internal/pipes"
+)
+
+// record pairs a value with the moment it arrived.
+type record[T any] struct {
+	val T
+	ts  time.Time
+}
+
+type slidingBatch[T any] struct {
+	items []record[T]
+	mu    sync.RWMutex
+}
+
+// NewSlidingBatch constructs a new sliding batch.
+func newSlidingBatch[T any]() *slidingBatch[T] {
+	return &slidingBatch[T]{
+		items: make([]record[T], 0),
+	}
+}
+
+// Push appends the item with its arrival timestamp.
+func (s *slidingBatch[T]) push(item T, now time.Time) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.items = append(s.items, record[T]{val: item, ts: now})
+}
+
+// next evicts old records and return the current window.
+func (s *slidingBatch[T]) next(
+	windowDuration time.Duration,
+	now time.Time,
+	final bool,
+) []T {
+	threshold := now.Add(-windowDuration)
+
+	s.mu.Lock()
+	// find cut index via binary search
+	cut := 0
+	for cut < len(s.items) && s.items[cut].ts.Before(threshold) {
+		cut++
+	}
+	if cut > 0 {
+		// reallocate so we don't hold old large backing arrays
+		tail := s.items[cut:]
+		newBuf := make([]record[T], len(tail))
+		copy(newBuf, tail)
+		s.items = newBuf
+	}
+
+	if len(s.items) == 0 {
+		s.mu.Unlock()
+		return []T{}
+	}
+
+	// snapshot values
+	window := make([]T, len(s.items))
+	for i, rec := range s.items {
+		window[i] = rec.val
+	}
+	if final {
+		s.items = nil
+	}
+
+	s.mu.Unlock()
+	// publish
+	return window
+}
```

Comment:
I think this code maybe O(n)

```go
	for cut < len(s.items) && s.items[cut].ts.Before(threshold) {
		cut++
	}
```

Consider this tweaked version

```suggestion
func (s *slidingBatch[T]) next(
	windowDuration time.Duration,
	now time.Time,
	final bool,
) []T {
	threshold := now.Add(-windowDuration)
	s.mu.Lock()
	// find cut index via binary search
	cut := sort.Search(len(s.items), func(i int) bool {
		return !s.items[i].ts.Before(threshold)
	})
	if cut > 0 {
		// reallocate so we don't hold old large backing arrays
		tail := s.items[cut:]
		newBuf := make([]record[T], len(tail))
		copy(newBuf, tail)
		s.items = newBuf
	}
	if len(s.items) == 0 {
		s.mu.Unlock()
		return []T{}
	}
	// snapshot values
	window := make([]T, len(s.items))
	for i, rec := range s.items {
		window[i] = rec.val
	}
	if final {
		s.items = nil
	}
	s.mu.Unlock()
	// publish
	return window
}
```

Reply:
For clarity, the only piece that changed was:

```
cut := sort.Search(len(s.items), func(i int) bool {
  return !s.items[i].ts.Before(threshold)
})
```

right?

Reply:
Correct